### ISSUE TITLE
Terminal resize roundtrip

### ISSUE THREAD

**Issue by jerch** [xtermjs/xterm.js#1914]
Issue:
Currently resizing the terminal control flow is done like this:
- `terminal.resize` triggered by some listener, which listens to grip movements (maybe with fit addon in between) or explicit col/row changes
- pty gets announced about size change by a sideway comm, this happens async and we dont wait for it to actually happen, basically we dont care when WINCH gets signalled
- buffers/views are resized (following a specific paradigm like reflow or simple cutting)

Since the pty resize happens async and we dont wait for it to happen there might be data in the buffers from pty to terminal that were written under the old size assumption. This is for most circumstances not a problem, most ASCII input flow scenarios will just work with this, but it might lead to garbage content lines if the slave application relies on the correct terminal size.

Solution:
I think this race condition cannot be avoided completely, but narrowed to only a small buffer window by doing the following:
- `terminal.resize` gets triggered by some listener
- terminal sends a XOFF to the pty, thus "halts" any outputting slave program
- terminal drains all pending data from buffers inserting it with old size
- terminal announces new size to pty and waits for resize ACK
- pty signals WINCH
- on resize ACK:
    - buffers/views are resized
    - XON to reenable data flow

Here the XOFF guarantees that pending data are already written and ACK token ensures that the pty has applied the new size. Have not yet thought about this model in detail, maybe the XOFF full stop could be made less invasive by inserting the ACK token in the websocket stream, maybe the ACK token should even contain the applied size (in case the pty cannot set the requested size for some reason). This size roundtrip certainly would slow down data flow during heavy slave output/terminal input.
Furthermore there still might be buffers involved that are not covered by XOFF and contain "wrongly sized" data, mainly on slave app side, like stdout FILE buffer. Not sure how libs like readline deal with buffered data on WINCH or if they always switch off any prebuffering.

Up for discussion, not sure if we should try to mess with this at all, for most scenarios it works good enough as it is now.

**Comment by Tyriar** [comment]
When I experimented with using XOFF it was a disaster, it breaks everything on Windows and zsh rebinds XOFF to some other hotkey.

**Comment by jerch** [comment]
@Tyriar Eww thats nasty. Wonder why a shell would mess with XON/XOFF at all. But doesnt this mean that our XON/XOFF based flow control in `terminal.write` does not work at all?

**Comment by Tyriar** [comment]
Yeah this stuff doesn't work, people should not use useFlowControl:

https://github.com/xtermjs/xterm.js/blob/cc531f19334f735b2aa0b0785afbfb0494a17e90/src/Terminal.ts#L1329-L1334

I think I left it there as it was default to off and we still want XON/XOFF to work when the user presses it, just not for xterm.js to start sending them. We can remove the useFlowControl related code if ctrl+s and ctrl+r still work in bash.

**Comment by jerch** [comment]
Not having any countermeasure against input data flooding is not advisable, worst case the interim buffers will grow up to JS engine limit. This might even kill the JS engine / xterm.js process. Hmm.

**Comment by jerch** [comment]
@Tyriar Ah good, so the direct sychronous invocation of the terminal input processing keeps the buffers from growing into nirvana. Thats pretty much like any C program would deal with real blocking buffers.

I think its still hazardous for remote usage scenarios which typically involve non blockable buffers (like websockets, ajax polling etc.) Maybe this is worth a browser specific terminal addon "flowcontrol" which can tell the pty interface (node-pty in the demo) to start/stop grabbing data (thus halting the slave application by buffer backpressure).

**Comment by Tyriar** [comment]
@jerch could be useful for attach, might be best for someone who uses this case to tackle that one as there are downsides to tweaking this (and the best numbers may depend on latency?)
